api/nodes/serializers.py

```diff
@@ -266,6 +269,104 @@ def update(self, node, validated_data):
+        node_settings = instance.get('settings', None)
+
+        try:
+            external_account_id = validated_data['settings']['external_account']['_id']
+            set_auth = True
+        except KeyError:
+            external_account_id = None
+            set_auth = False
+
+        try:
+            folder_id = validated_data['settings']['folder_id']
+            set_folder = True
+        except KeyError:
+            folder_id = None
+            set_folder = False
+
+        if folder_id and addon_name == 'googledrive':
+            try:
+                folder_path = validated_data['settings']['folder_path']
+            except KeyError:
+                folder_path = None
+            folder_id = {
+                'id': folder_id,
+                'path': folder_path
+            }
+
+        enabled = validated_data.get('enabled', None)
+
+        if enabled is False and (folder_id or external_account_id):
+            raise Conflict('Cannot disable the addon and set the folder/authorization at the same time.')
+
+        node = self.context['view'].get_node()
+        if ((not node_settings or not node_settings.has_auth)
+           and folder_id and not external_account_id):
+            raise Conflict('Cannot set folder without authorization')
+
+        if node_settings and enabled is False:
+            node.delete_addon(addon_name, auth)
+            node_settings = None
+            external_account_id = None
+            folder_id = None
+            enabled = False
+
+        if not node_settings and (enabled is not False or (folder_id or external_account_id)):
+            node_settings = node.get_or_add_addon(addon_name, auth=auth)
+
+        if node_settings and node_settings.configured and set_folder and not folder_id:
+            node_settings.clear_settings()
+            node_settings.save()
+
+        if node_settings and node_settings.has_auth and set_auth and not external_account_id:
+            node_settings.deauthorize(auth=auth)  # clear_auth performs save
+
+        elif external_account_id:
+            external_account = ExternalAccount.load(external_account_id)
+            if not external_account:
+                raise exceptions.NotFound('Unable to find requested account.')
+            if external_account not in auth.user.external_accounts:
+                raise exceptions.PermissionDenied('Requested action requires account ownership.')
+            if node_settings.external_account and external_account_id != node_settings.external_account._id:
+                node_settings.deauthorize(auth=auth)
+            node_settings.set_auth(external_account, auth.user)
+
+        if (folder_id
+            and not (instance['settings']
+                and instance['settings'].get('folder_id', False)
+                and (instance['settings']['folder_id'] == folder_id
+                    or instance['settings']['folder_id'] == folder_id.get('folder_id', False)
```

Comment:
so `folder_id` can be a str or a dict? 

I wish I had more constructive thoughts on this, but this block is pretty hard to grok.